Rework `/proc/.../maps` parsing

danielocfb opened the pull request:
Please see individual commits for descriptions.

anakryiko reviewed (approved):
Some notes on regex, as I was confused why it even works initially :) Doesn't have to be addressed right now, of course. LGTM, otherwise.

danielocfb commented:
We discussed the PR offline. The regex is problematic but it's going away. I reworked the changes to fix one bug that I noticed after discussion, which was handling of "xxx (deleted)" paths. I also added more tests for more `maps` formats I found (some of which were broken beforehand, as outlined in the description).

I also tried using `/proc/{pid}/map_files` unconditionally, but because the logic build on top of that is basically broken that won't work without out of the box. Will do that later, as it was not the objective of this pull request and it's not a new issue.
